fix: Fix combining of ClipPath child elements on Android

wodin opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please follow the template so that the reviewers can easily understand what the code changes affect -->

# Summary

Fixes combining of ClipPath child elements to match the spec and browsers, Inkscape, etc.
The problem affects both Android and iOS, but this PR currently only fixes it on Android
I started looking at the iOS code, but got a bit lost

Relevant issue: #1520

The code was conflating clipRules with how child elements need to be combined. These are unrelated. Children always need to be combined using UNION

From https://www.w3.org/TR/SVG11/masking.html#EstablishingANewClippingPath
> When the ‘clipPath’ element contains multiple child elements, the silhouettes of the child elements are logically OR'd together to create a single silhouette which is then used to restrict the region onto which paint can be applied.

### Related commits
766926f75874dedd7b13dadf65315f7030c4234a
a1097b85942c559ebcef6b93fa2ce601434d9c50

### Note

This is a breaking change and the ClipPath example image with need to be updated in USAGE.

## Test Plan

`yarn test` gave the following warning on the `main` branch (i.e. before my change. My fix is completely unrelated to this file):
```
/tmp/react-native-svg/src/css/LocalSvg.tsx
  26:34  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
```

After adding `// eslint-disable-next-line @typescript-eslint/no-explicit-any` on the previous line, everything is happy with my changes applied:

```console
% yarn test
yarn run v1.22.19
$ npm run lint && npm run tsc

> react-native-svg@14.1.0 lint
> eslint --ext .ts,.tsx src


> react-native-svg@14.1.0 tsc
> tsc --noEmit

✨  Done in 5.09s.
```

`yarn jest` failed before my change. I ran `yarn jest -u` on the `main` branch to update the snapshots.
After that, `yarn jest` passes on the branch containing my change.

```console
% git checkout wodin/fix-combining-of-clip-path-children
M	__tests__/__snapshots__/css.test.tsx.snap
M	src/css/LocalSvg.tsx
Switched to branch 'wodin/fix-combining-of-clip-path-children'
% yarn jest
yarn run v1.22.19
$ jest
 PASS  __tests__/css.test.tsx
  ✓ inlines styles (6 ms)
  ✓ supports CSS in style element (8 ms)

Test Suites: 1 passed, 1 total
Tests:       2 passed, 2 total
Snapshots:   2 passed, 2 total
Time:        0.737 s, estimated 1 s
Ran all test suites.
✨  Done in 2.72s.
```

### What's required for testing (prerequisites)?

N/A

### What are the steps to reproduce (after prerequisites)?

Code like this reproduces the problem and demonstrates that the fix works:
```javascript
    <Svg viewBox="0 0 250 234">
      <ClipPath id="clip">
        <Polygon points="31,0 31,234 170,234 250,0" />
        <Rect x="0" y="0" width="41" height="234" rx="8" ry="8" />
        <Rect x="0" y="112" width="250" height="10" />
      </ClipPath>
      <Rect x="0" y="0" width="250" height="234" fill="red" clipPath="url(#clip)" />
    </Svg>
```

Before the fix the result looks like this:
<img src="https://github.com/software-mansion/react-native-svg/assets/4579020/a42f8a10-9eab-45e0-ac14-abffb0c91af3" alt="Before" width="200px">

After the fix it looks like this:
<img src="https://github.com/software-mansion/react-native-svg/assets/4579020/3d74a3c5-9072-44e0-9a70-d8c056652eb4" alt="After" width="200px">

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ❌     |
| Android |    ✅     |

## Checklist

<!-- Check completed item, when applicable, via: [X] -->

- [X] I have tested this on a device and a simulator
- [ ] I added documentation in `README.md`
- [ ] I updated the typed files (typescript)
- [ ] I added a test for the API in the `__tests__` folder

wodin commented:
@WoLewicki I would love your thoughts on this if you have any?

Unfortunately I have no experience with the iOS side of this, so struggled to get anywhere with it, and have run out of energy for the moment.

While looking into this I came across the [SVG 1.1 Second Edition Test Suite](https://www.w3.org/Graphics/SVG/Test/20110816/)

I wonder whether it would be useful to take the test cases from the test suite and display the RNSVG version and a WebView version next to each other for visual comparison?

WoLewicki commented:
I am no expert in this code either unfortunately. As for displaying the web versions of the components next to the original ones, I am up for it 🚀 We also have a web version inside `Example` app so it can be also seen there. Do you want me to merge this one for start or to wait until the solution on `iOS` is also found to keep the behavior consistent?

wodin commented:
I'm not sure whether it's better to be consistent between Android and iOS or at least fix clipPath on Android

wodin commented:
> We also have a web version inside `Example` app so it can be also seen there.

Do you mean `react-native-web`? OK. Although I think it might be easier to compare them if they're on the same screen

I thought of maybe running `svgr` during the build process to convert the `.svg` files from the test suite to components.
I suppose we'd probably have to delete all the ones that contain animation? And other stuff like `<set>` and `<script>` etc.

wodin commented:
Here's one that's relevant to this PR. This is without the patch. I will add a screenshot using the patched version later:

[masking-path-01-b.svg](https://www.w3.org/Graphics/SVG/Test/20110816/harness/htmlSVGWeb/masking-path-01-b.html)
![masking-path-01-b.svg](https://github.com/software-mansion/react-native-svg/assets/4579020/6c7e9945-bcda-44e0-af34-af42f4aa43dd)

>  ## Pass Criteria
> 
> The test at the top shows an orange rectangle (with black stroke) being clipped by another rectangle. So only the middle portion of the orange rectangle should be visible. Also the black stroke should only be visible along the top and bottom edge of the rectangle.
> 
> The example at the bottom has a group containing a text string and two rectangles. The group has a clipping path defined using two overlapping rectangles. ***Of concern is the overlapping area shared by the two rectangles. There should not be holes in this overlapping area, the clip region is the union of the two rectangles.*** For clarity, guide rectangles in grey show the position of the clipping rectangles.
> 
> The rendered picture should match the reference image exactly, except for possible variations in the labelling text (per CSS2 rules).
> ## Test Description
> 
> Test to see if the basic clipping works using the `clipPath` element and the `clip-path` property.
> 
> This test uses the following elements : `<clipPath>` and the following properties : `clip-path`.
>

wodin commented:
OK @jakex7, I've made more updates for this and added relevant test cases from the W3C SVG test suite.
I have *not* yet been able to test properly on iOS/macOS. Having some space issues on my machine :-/
I also only tested with `paper-example`. Let me know if I should also test with `fabric-example` or other examples?

wodin commented:
Hah! Found a bug in the way the generate script renders this SVG:
### SVG
This renders correctly with my fixes and matches what I see in Firefox, Chrome and Inkscape

![clip-rule](https://github.com/user-attachments/assets/ae22563c-533a-42ed-b8f4-30a298304690)

### Reference image
This is the reference image generated by `yarn generateE2eReferences`

<img width="600" height="600" alt="clip-rule" src="https://github.com/user-attachments/assets/96a0d225-d8b6-447f-90be-ae7714bc81f4" />

wodin commented:
I've replaced the W3C tests with new ones and tidied up the branch, so no more test case license uncertainty.

The reference generation is still broken for `clip-rule.svg`, so I generated those manually using Inkscape, but I verified it looks the same in Chrome (v143) and Firefox as well. It's just the old Chrom v127 used by Puppeteer that's broken.

I also added a test case based on #1520. This one still fails on iOS after the changes because of antialiasing differences.

Other failed cases are unrelated to these changes:

**clip-path-shapes (iOS + Android):**
- `clipPathUnits="objectBoundingBox"` not supported - normalized 0-1 coordinates aren't being scaled to element's bounding box

**mask-edge-cases (iOS + Android):**
- `visibility:hidden` inside `<mask>` doesn't hide the element - mask content still renders

I tested on both PaperExample and FabricExample. I discovered that the macOS examples don't support the E2E tests.
